packages/react-dom/src/client/DOMPropertyOperations.js

```diff
@@ -125,84 +111,64 @@ export function getValueForAttribute(node, name, expected) {
  * @param {string} name
  * @param {*} value
  */
-export function setValueForProperty(node, name, value) {
-  const propertyInfo = getPropertyInfo(name);
-
-  if (propertyInfo && shouldSetAttribute(name, value)) {
-    if (shouldIgnoreValue(propertyInfo, value)) {
-      deleteValueForProperty(node, name);
-      return;
-    } else if (propertyInfo.mustUseProperty) {
-      // Contrary to `setAttribute`, object properties are properly
-      // `toString`ed by IE8/9.
-      node[propertyInfo.propertyName] = value;
-    } else {
-      const attributeName = propertyInfo.attributeName;
-      const namespace = propertyInfo.attributeNamespace;
-      // `setAttribute` with objects becomes only `[object]` in IE8/9,
-      // ('' + value) makes it output the correct toString()-value.
-      if (namespace) {
-        node.setAttributeNS(namespace, attributeName, '' + value);
-      } else if (
-        propertyInfo.hasBooleanValue ||
-        (propertyInfo.hasOverloadedBooleanValue && value === true)
-      ) {
-        node.setAttribute(attributeName, '');
+export function setValueForProperty(
+  node: Element,
+  name: string,
+  value: mixed,
+  isCustomComponentTag: boolean,
+) {
+  if (shouldSkipAttribute(name, isCustomComponentTag)) {
+    return;
+  }
+  const propertyInfo = isCustomComponentTag ? null : getPropertyInfo(name);
+  if (shouldTreatAttributeValueAsNull(name, value, isCustomComponentTag)) {
+    value = null;
+  }
+  // If the prop is in the special list, treat it as a simple attribute.
+  if (!propertyInfo) {
+    if (isAttributeNameSafe(name)) {
+      const attributeName = name;
+      if (value == null) {
+        node.removeAttribute(attributeName);
       } else {
-        node.setAttribute(attributeName, '' + value);
+        node.setAttribute(attributeName, '' + (value: any));
       }
     }
-  } else {
-    setValueForAttribute(
-      node,
-      name,
-      shouldSetAttribute(name, value) ? value : null,
-    );
     return;
   }
-}
-
-export function setValueForAttribute(node, name, value) {
-  if (!isAttributeNameSafe(name)) {
+  const {
+    hasBooleanValue,
+    hasOverloadedBooleanValue,
+    mustUseProperty,
+  } = propertyInfo;
+  if (mustUseProperty) {
+    const {propertyName} = propertyInfo;
+    if (value === null) {
+      (node: any)[propertyName] = hasBooleanValue ? false : '';
```

Comment:
All attributes that must use property are boolean values, so technically this check is unnecessary. We could add a new attribute type like HAS_BOOLEAN_PROPERTY to simplify this, and get rid of the bitmasks again (happy to do in a follow up)

Reply:
Let's follow up. This diff is getting hard to deal with.